plugins/chat/app/controllers/chat/api/direct_messages_controller.rb

```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+
+# TODO (martin) Remove this endpoint when we move to do the channel creation
+# when a message is first sent to avoid double-request round trips for DMs.
+class Chat::Api::DirectMessagesController < Chat::ApiController
+  def create
+    channel_params = params.require(:target_usernames)
+
+    with_service(Chat::CreateDirectMessageChannel, **channel_params) do
+      on_success do
+        render_serialized(
+          result.channel,
+          Chat::ChannelSerializer,
+          root: "channel",
+          membership: result.membership,
+        )
+      end
+      on_model_not_found(:target_users) { raise ActiveRecord::RecordNotFound }
+      on_failed_policy(:does_not_exceed_max_direct_message_users) do
+        error_message =
+          if SiteSetting.chat_max_direct_message_users == 0
+            I18n.t("chat.errors.over_chat_max_direct_message_users_allow_self")
+          else
+            I18n.t(
+              "chat.errors.over_chat_max_direct_message_users",
+              count: SiteSetting.chat_max_direct_message_users + 1, # +1 for the acting user
+            )
+          end
+
+        raise Discourse::InvalidParameters.new(:target_usernames, error_message)
+      end
+      on_failed_policy(:acting_user_not_disallowing_all_messages) do
+        render_json_error(I18n.t("chat.errors.actor_disallowed_dms"))
+      end
+      on_failed_policy(:acting_user_can_message_all_target_users) do
+        render_json_error(
+          I18n.t(
+            "chat.errors.not_accepting_dms",
+            username: result.preventing_communication_username,
+          ),
+        )
+      end
+      on_failed_policy(:acting_user_not_preventing_messages_from_any_target_users) do
+        render_json_error(
+          I18n.t(
+            "chat.errors.actor_preventing_target_user_from_dm",
+            username: result.preventing_communication_username,
+          ),
+        )
+      end
+      on_failed_policy(:acting_user_not_ignoring_any_target_users) do
+        render_json_error(
+          I18n.t(
+            "chat.errors.actor_ignoring_target_user",
+            username: result.preventing_communication_username,
+          ),
+        )
+      end
+      on_failed_policy(:acting_user_not_muting_any_target_users) do
+        render_json_error(
+          I18n.t(
+            "chat.errors.actor_muting_target_user",
+            username: result.preventing_communication_username,
+          ),
+        )
+      end
+      on_model_errors(:direct_message) do
+        render_json_error(result.direct_message, type: :record_invalid, status: 422)
+      end
+      on_model_errors(:channel) do
+        render_json_error(result.channel, type: :record_invalid, status: 422)
+      end
```

Comment:
@Flink what's your feeling on this ? How would you suggest we simplify it?

Reply:
I’m not sure to agree with the fact it’s unusable outside of controllers. Since we know which policy failed, we can act on it as we see fit. It’s indeed nicer/easier with models and contracts, since they are validation objects and thus contain detailed errors.

As for simplifying this, first, it doesn’t shock me that much since we want to display several error messages to the end-user, we have to match several policies. Now, I guess we could try different approaches:

1. create a small helper in the controller to provide the proper error message based on the policy name.
2. regroup all the polices into a single one, generate the error message and put it into the context.
3. allow `policy` to take a “real” policy object, allowing to store arbitrary data and expose this object in the resulting context.
4. something else :grin: 

I think 1) is relatively similar to what Joffrey suggested, having something inspecting the result and outputting the expected error. 2) is going back to what we had initially, but just put that into the service and make it work somehow. While I guess this would work, I think the result would be a bit clumsy.
3) would take a bit of extra work, but would probably be better in the long run. We could have self-contained policies, able to render themselves as errors if needed. It would normalize the way of doing things and those objects should be quite easy to extend. We could probably do something like this: `result["result.policy.my_policy"].reason` and it would be up to the policy to know how to output why it failed.

We may have to add some helpers at the controller level anyway (we could have `record_invalid :direct_message` for example, that would take care of calling `render_json_error` with the proper args, etc.).

I can start taking a look at what could be a small policy object and see how it goes :slightly_smiling_face: